### ISSUE TITLE
[Bug] [Android] Crash when calling Platform.ClearRenderer(...)

### ISSUE THREAD

**Issue by velocitysystems** [xamarin/Xamarin.Forms#11029]
### Description
A crash occurs when calling `Platform.ClearRenderer(..)` from the Dispose method block in a custom renderer.

_Previously custom renderers which needed to cleanup and detach the renderer from the `Element` would call `Platform.ClearRenderer()` in the dispose method block. This worked for several years up until recently._

### Steps to Reproduce
Create a custom renderer.
In the Dispose method block, call `Platform.ClearRenderer(..)` to detach the renderer from the Element.
App crash with stack trace below.

### Expected Behavior
The app should not crash.

### Actual Behavior
The app crashes with the following exception and stack trace:
**System.ArgumentException: Must have a valid JNI object reference!**

```
06-12 12:30:42.757 I/MonoDroid( 8577): UNHANDLED EXCEPTION:
06-12 12:30:42.763 I/MonoDroid( 8577): System.ArgumentException: Must have a valid JNI object reference!
06-12 12:30:42.763 I/MonoDroid( 8577): Parameter name: value
06-12 12:30:42.763 I/MonoDroid( 8577):   at Android.Runtime.AndroidValueManager.RemovePeer (Java.Interop.IJavaPeerable value) [0x0001e] in <55654ebe9f2a48e6bade2862bb243f94>:0 
06-12 12:30:42.763 I/MonoDroid( 8577):   at Java.Interop.JniRuntime+JniValueManager.DisposePeer (Java.Interop.JniObjectReference h, Java.Interop.IJavaPeerable value) [0x0001f] in <dac4c5a4b77f4e61a5e6d9d3050dfb9f>:0 
06-12 12:30:42.763 I/MonoDroid( 8577):   at Java.Interop.JniRuntime+JniValueManager.DisposePeer (Java.Interop.IJavaPeerable value) [0x00038] in <dac4c5a4b77f4e61a5e6d9d3050dfb9f>:0 
06-12 12:30:42.763 I/MonoDroid( 8577):   at Java.Lang.Object.Dispose () [0x00000] in <55654ebe9f2a48e6bade2862bb243f94>:0 
06-12 12:30:42.763 I/MonoDroid( 8577):   at Xamarin.Forms.Platform.Android.VisualElementPackager.RemoveChild (Xamarin.Forms.VisualElement view) [0x0007f] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementPackager.cs:275 
06-12 12:30:42.763 I/MonoDroid( 8577):   at Xamarin.Forms.Platform.Android.VisualElementPackager.SetElement (Xamarin.Forms.VisualElement oldElement, Xamarin.Forms.VisualElement newElement) [0x000ab] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementPackager.cs:321 
06-12 12:30:42.763 I/MonoDroid( 8577):   at Xamarin.Forms.Platform.Android.VisualElementPackager.Dispose (System.Boolean disposing) [0x000ad] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementPackager.cs:88 
06-12 12:30:42.764 I/MonoDroid( 8577):   at Xamarin.Forms.Platform.Android.VisualElementPackager.Dispose () [0x00000] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementPackager.cs:61 
06-12 12:30:42.764 I/MonoDroid( 8577):   at Xamarin.Forms.Platform.Android.VisualElementRenderer`1[TElement].Dispose (System.Boolean disposing) [0x00081] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementRenderer.cs:306 
06-12 12:30:42.764 I/MonoDroid( 8577):   at Xamarin.Forms.Platform.Android.Platform+DefaultRenderer.Dispose (System.Boolean disposing) [0x0001a] in D:\a\1\s\Xamarin.Forms.Platform.Android\Platform.cs:1306 
06-12 12:30:42.764 I/MonoDroid( 8577):   at Java.Lang.Object.Java.Interop.IJavaPeerable.Disposed () [0x00000] in <55654ebe9f2a48e6bade2862bb243f94>:0 
06-12 12:30:42.764 I/MonoDroid( 8577):   at Java.Interop.JniRuntime+JniValueManager.DisposePeer (Java.Interop.JniObjectReference h, Java.Interop.IJavaPeerable value) [0x00019] in <dac4c5a4b77f4e61a5e6d9d3050dfb9f>:0 
06-12 12:30:42.764 I/MonoDroid( 8577):   at Java.Interop.JniRuntime+JniValueManager.DisposePeer (Java.Interop.IJavaPeerable value) [0x00038] in <dac4c5a4b77f4e61a5e6d9d3050dfb9f>:0 
06-12 12:30:42.764 I/MonoDroid( 8577):   at Java.Lang.Object.Dispose () [0x00000] in <55654ebe9f2a48e6bade2862bb243f94>:0 
06-12 12:30:42.764 I/MonoDroid( 8577):   at Xamarin.Forms.Platform.Android.VisualElementPackager.RemoveChild (Xamarin.Forms.VisualElement view) [0x0007f] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementPackager.cs:275 
06-12 12:30:42.764 I/MonoDroid( 8577):   at Xamarin.Forms.Platform.Android.VisualElementPackager.SetElement (Xamarin.Forms.VisualElement oldElement, Xamarin.Forms.VisualElement newElement) [0x000ab] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementPackager.cs:321 
06-12 12:30:42.764 I/MonoDroid( 8577):   at Xamarin.Forms.Platform.Android.VisualElementPackager.Dispose (System.Boolean disposing) [0x000ad] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementPackager.cs:88 
06-12 12:30:42.765 I/MonoDroid( 8577):   at Xamarin.Forms.Platform.Android.VisualElementPackager.Dispose () [0x00000] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementPackager.cs:61 
06-12 12:30:42.765 I/MonoDroid( 8577):   at Xamarin.Forms.Platform.Android.VisualElementRenderer`1[TElement].Dispose (System.Boolean disposing) [0x00081] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementRenderer.cs:306 
06-12 12:30:42.765 I/MonoDroid( 8577):   at Xamarin.Forms.Platform.Android.PageRenderer.Dispose (System.Boolean disposing) [0x00024] in D:\a\1\s\Xamarin.Forms.Platform.Android\Renderers\PageRenderer.cs:59 
06-12 12:30:42.765 I/MonoDroid( 8577):   at Java.Lang.Object.Java.Interop.IJavaPeerable.Disposed () [0x00000] in <55654ebe9f2a48e6bade2862bb243f94>:0 
06-12 12:30:42.765 I/MonoDroid( 8577):   at Java.Interop.JniRuntime+JniValueManager.DisposePeer (Java.Interop.JniObjectReference h, Java.Interop.IJavaPeerable value) [0x00019] in <dac4c5a4b77f4e61a5e6d9d3050dfb9f>:0 
06-12 12:30:42.766 I/MonoDroid( 8577):   at Java.Interop.JniRuntime+JniValueManager.DisposePeer (Java.Interop.IJavaPeerable value) [0x00038] in <dac4c5a4b77f4e61a5e6d9d3050dfb9f>:0 
06-12 12:30:42.766 I/MonoDroid( 8577):   at Java.Lang.Object.Dispose () [0x00000] in <55654ebe9f2a48e6bade2862bb243f94>:0 
06-12 12:30:42.766 I/MonoDroid( 8577):   at Xamarin.Forms.Platform.Android.AppCompat.FragmentContainer.OnDestroyView () [0x0003c] in D:\a\1\s\Xamarin.Forms.Platform.Android\AppCompat\FragmentContainer.cs:88 
06-12 12:30:42.766 I/MonoDroid( 8577):   at Android.Support.V4.App.Fragment.n_OnDestroyView (System.IntPtr jnienv, System.IntPtr native__this) [0x00008] in <fefee6c2c695459088a9df092723e052>:0 
06-12 12:30:42.766 I/MonoDroid( 8577):   at (wrapper dynamic-method) Android.Runtime.DynamicMethodNameCounter.68(intptr,intptr) in D:\a\1\s\Xamarin.Forms.Platform.And
```

### Basic Information
- Version with issue: 4.6.0-sr2 and above
- Last known good version: N/A - Seems related to Xamarin.Android/Mono itself.
- IDE: Microsoft Visual Studio 2019 Enterprise 16.6.2
- Platform Target Frameworks: 
  - Android SDK: 10.3.1.4
- Android Support Library Version: 28.0.0.3

### Reproduction Link
To be supplied.

### Workaround
See workaround documented below.

**Comment by velocitysystems** [comment]
Testing demonstrates this may be related to this [commit](https://github.com/xamarin/Xamarin.Forms/commit/e87323b9327eaa9374342782be016ef86d38995b), and I believed this is confined to a custom renderer we are using.

@samhouts @hartez Follow the changes made in the commit, do you have any recommendations for changes that need to be made to custom renderers to make them compatible with recent changes made?

**Comment by velocitysystems** [comment]
@samhouts @hartez Upon further examining the changes, we have solved this issue in our custom renderer(s) by adding the following lines to the `OnElementPropertyChanged` handler:

```
if (this.IsDisposed())
{
    return;
}
```

Can you confirm this is now a requirement for custom renderers? Unfortunately this change caught us out by surprise as it wasn't called out in the XF release notes.

**Comment by hartez** [comment]
> We are experiencing this in a custom fast renderer we have implemented.
> Testing demonstrates this may be related to this [commit](https://github.com/xamarin/Xamarin.Forms/commit/e87323b9327eaa9374342782be016ef86d38995b) and/or [commit](https://github.com/xamarin/Xamarin.Forms/commit/8fb8ea36675c713866be828c6d57a3dc3cc59fad#diff-3f320618f3c51c03d04d1ec79ca025ef).
> 
> @samhouts @hartez Follow the changes made in the commit(s), do you have any recommendations for changes that need to be made to custom renderers to make them compatible with recent changes made?

The changes in e87323b9327eaa9374342782be016ef86d38995b don't create any _new_ requirements for a custom renderer, they are just a guard against an issue with renderers in Android which has always been there. I suppose I'd say that it's a good idea to add that check to any override of `OnElementPropertyChanged` you implement in a custom renderer. Hopefully in a future version we can find a way to make those checks redundant.

As for 8fb8ea36675c713866be828c6d57a3dc3cc59fad, that's just correcting an improper implementation of the DIspose pattern in one class; it should have no effect on how anyone uses the class or writes any custom renderers. If that commit _does_ turn out to be the one that caused your issue, I'd be very keen to understand why. 

Why did you have to stop using `Platform.ClearRenderer()`? I'm not sure I understand that change.

**Comment by velocitysystems** [comment]
Thanks @hartez.

The reason we stopped using `Platform.ClearRenderer()` was that this was the source of the JNI crash reported in the ticket description. We noticed that none of the stock XF fast renderers use this method, but rather simply clear the value of the attached `Renderer` property.

Looking at the implementation in the `Platform` class, there is _additional_ behaviour in this method as opposed to simply clearing out the attached property. Possibly this is conflicting with the changes made in 8fb8ea3.

https://github.com/xamarin/Xamarin.Forms/blob/bd6ef179c95042403930e08d0096e24f1d1c42a4/Xamarin.Forms.Platform.Android/Platform.cs#L320-L335

In our case, using the workaround we have documented above and aligning our custom fast renderer exactly with the pattern used by the stock XF renderers resolved the issue for us.

**Comment by hartez** [comment]
@velocitysystems Do you know which line of ClearRenderer() was throwing the exception from the original stack trace?

**Comment by velocitysystems** [comment]
@hartez My apologies for the delay in replying, have been so snowed under with an upcoming release.
Further testing seems to indicate this is not so much tied to XF 4.6.0, but possibly to an underlying change in Mono/Xamarin.Android.

We recently had to release a patch release which is still running XF 4.40-sr2 and saw this issue when compiling the same code base with the latest tooling. Similarly, we had a custom renderer which was using `Platform.ClearRenderer()` in the `Dispose()` block. Changing this to workaround (above) resolved the issue.

**Repro**
Tabbed page with a `BottomNavigationView`.
Page 1 is a blank content page.
Page 2 has a custom renderer based on a `RecyclerView`.
Switch tabs back and forth, crash.

All our stack traces aren't particularly helpful, here is a typical one....

**AndroidValueManager.RemovePeer (Java.Interop.IJavaPeerable value)
System.ArgumentException: Must have a valid JNI object reference! Parameter name: value**

```
AndroidValueManager.RemovePeer (Java.Interop.IJavaPeerable value)
JniRuntime+JniValueManager.DisposePeer (Java.Interop.JniObjectReference h, Java.Interop.IJavaPeerable value)
JniRuntime+JniValueManager.DisposePeer (Java.Interop.IJavaPeerable value)
Object.Dispose ()
VisualElementRenderer`1[TElement].Dispose (System.Boolean disposing)
Platform+DefaultRenderer.Dispose (System.Boolean disposing)
IJavaPeerable.Disposed ()
JniRuntime+JniValueManager.DisposePeer (Java.Interop.JniObjectReference h, Java.Interop.IJavaPeerable value)
JniRuntime+JniValueManager.DisposePeer (Java.Interop.IJavaPeerable value)
Object.Dispose ()
VisualElementRenderer`1[TElement].Dispose (System.Boolean disposing)
Platform+DefaultRenderer.Dispose (System.Boolean disposing)
IJavaPeerable.Disposed ()
JniRuntime+JniValueManager.DisposePeer (Java.Interop.JniObjectReference h, Java.Interop.IJavaPeerable value)
JniRuntime+JniValueManager.DisposePeer (Java.Interop.IJavaPeerable value)
Object.Dispose ()
VisualElementRenderer`1[TElement].Dispose (System.Boolean disposing)
PageRenderer.Dispose (System.Boolean disposing)
IJavaPeerable.Disposed ()
JniRuntime+JniValueManager.DisposePeer (Java.Interop.JniObjectReference h, Java.Interop.IJavaPeerable value)
JniRuntime+JniValueManager.DisposePeer (Java.Interop.IJavaPeerable value)
Object.Dispose ()
FragmentContainer.OnDestroyView ()
Fragment.n_OnDestroyView (System.IntPtr jnienv, System.IntPtr native__this)
(wrapper dynamic-method) Android.Runtime.DynamicMethodNameCounter.65(intptr,intptr)
```

When the fragment is destroyed, it naturally calls the Dispose() method on the custom renderer and the crash originates at the call to `Platform.ClearRenderer().`

Looking at the code in AndroidRuntime.cs, I believe it is blowing up [here](https://github.com/xamarin/xamarin-android/blob/d471b4a4308310afb8fa3fa1047003a7caedd57f/src/Mono.Android/Android.Runtime/AndroidRuntime.cs#L557-L568):

```
public override void RemovePeer (IJavaPeerable value)
{
    if (value == null)
        throw new ArgumentNullException (nameof (value));
    if (!value.PeerReference.IsValid)
         throw new ArgumentException ("Must have a valid JNI object reference!", nameof (value));
    var reference = value.PeerReference;
    var hash = JNIEnv.IdentityHash! (reference.Handle);
    RemovePeer (value, hash);
}
```

The bottom line is `PeerReference.IsValid` must be returning false due to some behavior in `Platform.ClearRenderer()`.
@hartez Since none of the XF renderers use this method, is it essentially considered deprecated then? Perhaps the API should be obsoleted?

As a workable solution, if `RendererProperty` was made public then any custom renderers could then implement the same method used by internal XF renderers and documented in the workaround above?

**Comment by velocitysystems** [comment]
@hartez @samhouts I've updated the ticket and documentation to more accurately reflect the actual bug.
Repro steps are included.

**Comment by hartez** [comment]
@velocitysystems Can you post the Dispose override with the call to `ClearRenderer()`? (The original one, not the new version using reflection.)

**Comment by velocitysystems** [comment]
Sure @hartez, this is for a custom fast renderer we have on Android for a media player.
We modelled the original design on the Xamarin.Forms fast-renderers and has been working fine for 2 years+.

**Original**
```
/// <summary>
/// Handle disposing and cleanup of the <see cref="Player"/>.
/// </summary>
/// <param name="disposing">True if can dispose managed resources, else false.</param>
protected override void Dispose(bool disposing)
{
    if (_disposed)
    {
        return;
    }

    _disposed = true;

    if (disposing)
    {
        Tracker?.Dispose();

        if (Element != null)
        {
            TearDownOldElement(Element as Player);

            if (Platform.GetRenderer(Element) == this)
            {
                Platform.ClearRenderer(this);
            }
        }
    }

    base.Dispose(disposing);
}
```

**Modified with Workaround**
```
/// <summary>
/// Handle disposing and cleanup of the <see cref="Player"/>.
/// </summary>
/// <param name="disposing">True if can dispose managed resources, else false.</param>
protected override void Dispose(bool disposing)
{
    if (_disposed)
    {
        return;
    }

    _disposed = true;

    if (disposing)
    {
        Tracker?.Dispose();

        if (Element != null)
        {
            TearDownOldElement(Element as Player);

            if (Platform.GetRenderer(Element) == this)
            {
                Element.ClearRenderer();
            }
        }
    }

    base.Dispose(disposing);
}
```

**Comment by PureWeen** [comment]
@velocitysystems without a repro this one is tricky to pin down

in your notes you say that sr2+ see the issue

Here's a compare of 4.6 rc with 4.6 sr2

https://github.com/xamarin/Xamarin.Forms/compare/release-4.6.0-rc2...release-4.6.0-sr3.1

I don't see anything in there that jumps out at me that would cause a difference in renderer life cycles

Though in your comments you reference commits that were tagged in the pre versions so maybe you meant pre2?

 I saw basically the same things @hartez saw when looking at those commits

I ran a few tests trying to recreate the JNI exception and from what I can tell this type of exception happens when the dispose ends up calling itself again

If I have the following code
```C#
renderer.Dispose();
renderer.Dispose();
```

That will run fine and the second dispose won't even call the dispose on the renderer


but if from inside the Dispose call somehow the dispose gets called again on the same class then it throws that JNI exception. So, from the stack trace I'm going to assume that each of these dispose calls here are being called onto the same class

```C#
06-12 12:30:42.763 I/MonoDroid( 8577):   at Xamarin.Forms.Platform.Android.VisualElementPackager.RemoveChild (Xamarin.Forms.VisualElement view) [0x0007f] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementPackager.cs:275 
06-12 12:30:42.763 I/MonoDroid( 8577):   at Xamarin.Forms.Platform.Android.VisualElementPackager.SetElement (Xamarin.Forms.VisualElement oldElement, Xamarin.Forms.VisualElement newElement) [0x000ab] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementPackager.cs:321 
06-12 12:30:42.763 I/MonoDroid( 8577):   at Xamarin.Forms.Platform.Android.VisualElementPackager.Dispose (System.Boolean disposing) [0x000ad] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementPackager.cs:88 
06-12 12:30:42.764 I/MonoDroid( 8577):   at Xamarin.Forms.Platform.Android.VisualElementPackager.Dispose () [0x00000] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementPackager.cs:61 
06-12 12:30:42.764 I/MonoDroid( 8577):   at Xamarin.Forms.Platform.Android.VisualElementRenderer`1[TElement].Dispose (System.Boolean disposing) [0x00081] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementRenderer.cs:306 
06-12 12:30:42.764 I/MonoDroid( 8577):   at Xamarin.Forms.Platform.Android.Platform+DefaultRenderer.Dispose (System.Boolean disposing) [0x0001a] in D:\a\1\s\Xamarin.Forms.Platform.Android\Platform.cs:1306 
06-12 12:30:42.764 I/MonoDroid( 8577):   at Java.Lang.Object.Java.Interop.IJavaPeerable.Disposed () [0x00000] in <55654ebe9f2a48e6bade2862bb243f94>:0 
06-12 12:30:42.764 I/MonoDroid( 8577):   at Java.Interop.JniRuntime+JniValueManager.DisposePeer (Java.Interop.JniObjectReference h, Java.Interop.IJavaPeerable value) [0x00019] in <dac4c5a4b77f4e61a5e6d9d3050dfb9f>:0 
06-12 12:30:42.764 I/MonoDroid( 8577):   at Java.Interop.JniRuntime+JniValueManager.DisposePeer (Java.Interop.IJavaPeerable value) [0x00038] in <dac4c5a4b77f4e61a5e6d9d3050dfb9f>:0 
06-12 12:30:42.764 I/MonoDroid( 8577):   at Java.Lang.Object.Dispose () [0x00000] in <55654ebe9f2a48e6bade2862bb243f94>:0 
06-12 12:30:42.764 I/MonoDroid( 8577):   at Xamarin.Forms.Platform.Android.VisualElementPackager.RemoveChild (Xamarin.Forms.VisualElement view) [0x0007f] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementPackager.cs:275 
06-12 12:30:42.764 I/MonoDroid( 8577):   at Xamarin.Forms.Platform.Android.VisualElementPackager.SetElement (Xamarin.Forms.VisualElement oldElement, Xamarin.Forms.VisualElement newElement) [0x000ab] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementPackager.cs:321 
```

I don't think any of this info is particularly useful to you here but it's what I can extract from the info given.

If you just don't call this code at all do you see any thing leaking? 

```C#
           if (Platform.GetRenderer(Element) == this)
            {
                Element.ClearRenderer();
            }
```

I feel like if you just leave this to the Framework the framework will just remove the rendererer and dispose of it

ClearRenderer is going to call dispose on that renderer as well so if you call ClearRenderer from dispose you're going to get a crash. 

I don't see anything on the XF side that would have caused this but maybe from the XA side?
Is there a particular XF version you can down grade to and then you don't see this issue?

**Comment by PureWeen** [comment]
https://github.com/xamarin/xamarin-android/issues/4989
